### PR TITLE
MAINT:sparse.linalg: Use _NoValue for deprecated kwargs

### PR DESCRIPTION
**Pull request by ilayn**
#### Reference issue
Continuation of #18488, as per #18624

#### What does this implement/fix?
Adds deprecation `_NoValue` check mechanism to iterative solvers in order to deprecate the `tol`  keyword.

### CODE REVIEW

File: scipy/sparse/linalg/_isolve/iterative.py [hunk 1]
```diff
@@ -567,7 +568,7 @@ def cgs(A, b, x0=None, tol=None, maxiter=None, M=None, callback=None,
         return postprocess(x), maxiter
 
 
-def gmres(A, b, x0=None, tol=None, restart=None, maxiter=None, M=None,
+def gmres(A, b, x0=None, tol=_NoValue, restart=None, maxiter=None, M=None,
```
Comment (j-bowhay) [inline scipy/sparse/linalg/_isolve/iterative.py:571]: I believe the `restrt` will also need bumping to 1.14.0 as well and for completeness we may as well change its default `_NoValue` too

File: scipy/sparse/linalg/_isolve/iterative.py [hunk 2]
```diff
@@ -3,15 +3,16 @@
...
-def _get_atol(name, b, tol=None, atol=0., rtol=1e-5):
+def _get_atol(name, b, tol=_NoValue, atol=0., rtol=1e-5):
     """
     A helper function to handle tolerance deprecations and normalization
     """
-    if tol is not None:
+    if tol is not _NoValue:
         msg = (f"'scipy.sparse.linalg.{name}' keyword argument 'tol' is "
                "deprecated in favor of 'rtol' and will be removed in SciPy "
                "v.1.14.0. Until then, if set, will override 'rtol'.")
```
Comment (h-vetinari) [inline scipy/sparse/linalg/_isolve/iterative.py:18]: Suggestion for drive-by typo fix
```suggestion
               "v.1.14.0. Until then, if set, it will override 'rtol'.")
```

File: scipy/sparse/linalg/_isolve/iterative.py (hunk 1 above)
Comment (h-vetinari) [reply scipy/sparse/linalg/_isolve/iterative.py:571]: Agreed, though for me that doesn't _necessarily_ fall in the purview of https://github.com/scipy/scipy/pull/18488 or this PR.

**Review by h-vetinari — COMMENTED** [review]
LGTM, modulo a small nit. Thanks!

File: scipy/sparse/linalg/_isolve/iterative.py (hunk 1 above)
Comment (ilayn) [reply scipy/sparse/linalg/_isolve/iterative.py:571]: While we are at it, indeed I've added to the deprecation cycle.

**Comment by ilayn** [comment]
Ah restrt turned a few warnings on, is everyone OK if I include those test filters here?

File: scipy/sparse/linalg/_isolve/iterative.py [hunk 3]
```diff
@@ -670,14 +671,14 @@ def gmres(A, b, x0=None, tol=None, restart=None, maxiter=None, M=None,
...
                          " keywords. Also 'rstrt' is deprecated."
-                         " and will be removed in SciPy 1.12.0. Use "
+                         " and will be removed in SciPy 1.14.0. Use "
                          "'restart' instad.")
-    if restrt is not None:
+    if restrt is not _NoValue:
         msg = ("'gmres' keyword argument 'restrt' is deprecated "
                "in favor of 'restart' and will be removed in SciPy"
-               " 1.12.0. Until then, if set, 'rstrt' will override 'restart'."
+               " 1.14.0. Until then, if set, 'rstrt' will override 'restart'."
```
Comment (j-bowhay) [inline scipy/sparse/linalg/_isolve/iterative.py:681]: Version in the docstring also needs bumping

File: scipy/sparse/linalg/_isolve/iterative.py [hunk 4]
```diff
@@ -670,14 +671,14 @@ def gmres(A, b, x0=None, tol=None, restart=None, maxiter=None, M=None,
 
     # Handle the deprecation frenzy
     if restrt and restart:
```
Comment (j-bowhay) [inline scipy/sparse/linalg/_isolve/iterative.py:673]: Test failures are a result of a bug here. `_NoValue` evaluates trues so we always incorrectly enter this block when `restart` is provided.
```suggestion
    if restrt not in (None, _NoValue) and restart:
```

File: scipy/sparse/linalg/_isolve/iterative.py (hunk 3 above)
Comment (ilayn) [reply scipy/sparse/linalg/_isolve/iterative.py:681]: One of these days, I'll start with coffee first before touching deprecations.

File: scipy/sparse/linalg/_isolve/iterative.py (hunk 4 above)
Comment (ilayn) [reply scipy/sparse/linalg/_isolve/iterative.py]: Nice. Thanks for catching that.

File: scipy/sparse/linalg/_isolve/iterative.py [hunk 5]
```diff
@@ -3,18 +3,19 @@
...
     A helper function to handle tolerance deprecations and normalization
     """
-    if tol is not None:
+    if tol is not _NoValue:
         msg = (f"'scipy.sparse.linalg.{name}' keyword argument 'tol' is "
                "deprecated in favor of 'rtol' and will be removed in SciPy "
-               "v.1.14.0. Until then, if set, will override 'rtol'.")
+               "v.1.14.0. Until then, if set, it will override 'rtol'.")
         warnings.warn(msg, category=DeprecationWarning, stacklevel=4)
         rtol = float(tol)
```
Comment (j-bowhay) [inline scipy/sparse/linalg/_isolve/iterative.py:20]: I think this needs a minor tweak too as I could call `bicg(A, b, None, None,...........)` and this would fail trying to cast None to a float.
```suggestion
        rtol = float(tol) if tol is not None else rtol
```

Comment (ilayn) [reply scipy/sparse/linalg/_isolve/iterative.py:20]: Yes but before that you will get a warning right? I think the context would be obvious if you ask me but I can add it too, I just don't want to promote the float setting to tol if the error clears

Comment (ilayn) [reply scipy/sparse/linalg/_isolve/iterative.py:20]: Actually, this would indeed fail unexpectedly so let's get it in anyways

**Review by j-bowhay — APPROVED** [review]
Thanks @ilayn LGTM. @h-vetinari did you want to have another look?

**Review by h-vetinari — APPROVED** [review]
Thanks!
